### PR TITLE
nautilus: mgr/dashboard: iSCSI targets not available if any gateway is down (and more...)

### PR DESCRIPTION
**Pull request by ricardoasmarques**
This backport PR contains the backport of 3 different issues.

1 - iSCSI targets not available if any gateway is down
----

backport tracker: https://tracker.ceph.com/issues/43333

backport of https://github.com/ceph/ceph/pull/31819
parent tracker: https://tracker.ceph.com/issues/42687

---

2 - Prevent deletion of iSCSI IQNs with open sessions
----

backport tracker: https://tracker.ceph.com/issues/43354

backport of https://github.com/ceph/ceph/pull/29133
parent tracker: https://tracker.ceph.com/issues/38332

---

3 - Javascript error when deleting an iSCSI target
----

backport tracker: https://tracker.ceph.com/issues/43355

backport of https://github.com/ceph/ceph/pull/30270
parent tracker: https://tracker.ceph.com/issues/41727

---

4 - Unable to remove an iSCSI gateway that is already in use
----

backport tracker: https://tracker.ceph.com/issues/43944

backport of https://github.com/ceph/ceph/pull/32984
parent tracker: https://tracker.ceph.com/issues/43900

---

this backport was staged using https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

**Comment by ricardoasmarques** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
```
119/172 Test   #5: run-tox-mgr-dashboard ...................***Failed  179.20 sec
```

**Comment by callithea** [comment]
jenkins test dashboard backend

**Comment by dillaman** [comment]
(rebased after nautilus branch whackiness)

**Comment by dillaman** [comment]
lgtm -- but it might make sense to also pull in #26891 so that you can have clean cherry-picks on the last commit.

**Comment by ricardoasmarques** [comment]
> lgtm -- but it might make sense to also pull in #26891 so that you can have clean cherry-picks on the last commit.

@dillaman Cherry-picking #26891 will cause even more conflicts, that's why I've decided to cut the chain here, instead of going deeper and deeper.

**Comment by ricardoasmarques** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/controllers/iscsi.py [hunk 1]
```diff
@@ -876,3 +902,31 @@ def _get_portals_by_host(portals):
...
+                             code='no_gateways_available',
+                             component='iscsi')
+
+
+def validate_rest_api(gateways):
+    for gateway in gateways:
+        try:
+            IscsiClient.instance(gateway_name=gateway).ping()
+        except RequestException:
+            raise DashboardException(msg='iSCSI REST Api not available for gateway '
```
Comment (votdev) [inline src/pybind/mgr/dashboard/controllers/iscsi.py]: s/Api/API/

**Comment by callithea** [comment]
- QA run [WIP]: http://pulpito.ceph.com/laura-2020-01-27_16:19:37-rados:dashboard-wip-laura-testing-nautilus-batch-8-distro-basic-mira/

**Comment by ricardoasmarques** [comment]
The following backport is now included in this backport PR:

- nautilus: mgr/dashboard: Unable to remove an iSCSI gateway that is already in use
  https://tracker.ceph.com/issues/43944

**Comment by callithea** [comment]
- [rados/mgr](http://pulpito.ceph.com/laura-2020-02-04_17:10:38-rados:mgr-wip-laura-testing-dashboard-nautilus-batch-8-distro-basic-smithi/) [WIP]
- [rados/dashboard](http://pulpito.ceph.com/laura-2020-02-04_17:10:06-rados:dashboard-wip-laura-testing-dashboard-nautilus-batch-8-distro-basic-smithi/) [WIP]

**Pull request merged**